Feat: use puppeteer for pdf rendering

jniles opened the pull request:
Removes the dependency on WKHTMLTOPDF and replaces it with puppeteer.  This supercedes https://github.com/IMA-WorldHealth/bhima/pull/3388 and is a lot less complex because we rely on an external library to inline assets.  Hopefully this will also work reasonably well across Windows and Linux.

If tests pass, we'll look into updating the documentation and updating styles.  Currently the rendered PDF looks like this on my machine:

![image](https://user-images.githubusercontent.com/896472/68322130-28b82180-00c3-11ea-9349-958f1ddc4c67.png)


Closes #3388.  Closes #3274.

-----

**Updates**
I've taken the opportunity to run through the major PDF reports used on the daily basis in BHIMA.  This is what they look like on my machine:

![tZxTMenHNb-min](https://user-images.githubusercontent.com/896472/68665964-4d891a80-0543-11ea-9893-c79f9d56c612.gif)

The following sizes work correctly:
 1. Thermal (210mm x 80mm)
 2. A6
 3. A4

I've updated the installation documentation to reflect the lack of WKHTMLTOPDF.  I've also removed wkhtmltopdf from the CI configuration.  We'll need to update https://github.com/IMA-WorldHealth/bhima/pull/3987 to reflect the lack of wkhtmltopdf after it lands.

------

My proposal is that we should wait to merge this until after we've released 1.7.0 (#3986).  That way, we are sure we have a stable base to roll back onto.

jniles committed:
chore: inline assets before puppeteer migration

To migrate the puppeteer for PDF rendering, we need to reliably be able
to inline assets (since puppeteer does not load them off disk like
wkhtmltopdf does).  Here, we've used inline-assets to post-process the
HTML rendered before rendering the PDF.

jniles committed:
feat(pdf): use puppeteer for PDF rendering

Removes WKHTMLTOPDF and replaces it with puppeteer for PDF rendering.
The styling needs to be improved, but the barebones is here.

Closes #3388.  Closes #3274.

mbayopanda reviewed (commented):
This feature requires to build lzma module, on Windows the lzma module is built correctly with Node 10 not with Node 13.

With Node10, the feature works fine there are some margin issues but in general the PDF is generated correctly

jniles committed:
chore: update receipt css for pdfs

We've made a bunch of changes to our rendering engines.  We front-load
the asset inlining to the HTML renderer, since it permits us to use the
same paths everywhere instead of maintaining local rendering paths (full
paths on disk) and HTTP-accessible paths.

To properly render out POS receipts, a new stylesheet has been created
and put into a handlebars partial.  We have a single, central location
to alter the styles of our POS receipts.  This also allowed us to remove
the underline partial since it can be done in a stylesheet much more
easily.

The dimensions of the documents are now defined in CSS rather than in
the PDF renderer.  The only options passed in are "preferCSSPageSize" to
ensure that the templates are rendered according to their internal
stylesheets.

jniles committed:
deps: bump deps locks to avoid juice errors

jniles committed:
chore(reports): migrate ReportManager to async fs

Migrates the promise-based backend to a fully async backend powered by
mz/fs.  Greatly improves code readability.

jniles committed:
chore(reports): wrap up PDF rendering w/ chrome

Puts the final touches on the PDF rendering with puppeteer, ensuring
that all POS, A6 and A4 pages render correctly.

jniles committed:
chore: rm references to wkhtmltopdf

Removes all references to wkhtmltopdf since we no longer use it.

jniles committed:
fix(pdf): remove wkhtmltopdf from tests

Removes the tests asserting that wkhtmltopdf is installed.  It is no
longer, since we use puppeteer.

jniles committed:
chore: use fs.promises instead of mz/fs

Future-proofs the ReportManager by using the fs.promises API stabilized
in Node 10.

mbayopanda commented on server/lib/ReportManager.js:
```diff
@@ -223,16 +220,10 @@ class ReportManager {
-        .catch(dfd.reject)
-        .done();
-    });
+    await fs.promises.writeFile(link, stream);
```
👍 great this promise version

mbayopanda reviewed (commented):
LGTM 👍

jniles committed:
Merge remote-tracking branch 'upstream/master' into feat-use-puppeteer-for-pdf-rendering

jniles committed:
Merge remote-tracking branch 'upstream/master' into feat-use-puppeteer-for-pdf-rendering
